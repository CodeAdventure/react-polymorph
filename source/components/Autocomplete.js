// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';


// external libraries
import createRef from 'create-react-ref/lib/createRef';
import _ from 'lodash';

// interal components
import { GlobalListeners } from './HOC/GlobalListeners';
import { withTheme } from './HOC/withTheme';

// internal utility functions
import { composeTheme, addThemeId } from '../utils/themes';
import { composeFunctions } from '../utils/props';

import { IDENTIFIERS } from '../themes/API';

type Props = {
  className: string,
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  error: string,
  invalidCharsRegex: RegExp,
  isOpeningUpward: boolean,
  label: string | Element<any>,
  maxSelections: number,
  maxVisibleOptions: number,
  multipleSameSelections: boolean,
  onChange: Function,
  options: Array<any>,
  preselectedOptions: Array<any>,
  placeholder: string,
  renderSelections: Function,
  renderOptions: Function,
  skin: ComponentType<any>,
  sortAlphabetically: boolean,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object
};

type State = {
  inputValue: string,
  error: string,
  selectedOptions: Array<any>,
  filteredOptions: Array<any>,
  isOpen: boolean,
  composedTheme: Object
};

class AutocompleteBase extends Component<Props, State> {
  // declare ref types
  rootElement: ?Element<any>;
  inputElement: ?Element<'input'>;
  suggestionsElement: ?Element<any>;
  optionsElement: ?Element<any>;

  // define static properties
  static displayName = 'Autocomplete';
  static defaultProps = {
    error: null,
    invalidCharsRegex: /[^a-zA-Z0-9]/g, // only allow letters and numbers by default
    isOpeningUpward: false,
    maxVisibleOptions: 10, // max number of visible options
    multipleSameSelections: true, // if true then same word can be selected multiple times
    options: [],
    sortAlphabetically: true, // options are sorted alphabetically by default
    theme: null,
    themeId: IDENTIFIERS.AUTOCOMPLETE,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    // define refs
    this.rootElement = createRef();
    this.inputElement = createRef();
    this.suggestionsElement = createRef();
    this.optionsElement = createRef();

    const {
      context,
      themeId,
      theme,
      themeOverrides,
      sortAlphabetically,
      options,
      preselectedOptions
    } = props;

    this.state = {
      inputValue: '',
      error: '',
      selectedOptions: preselectedOptions || [],
      filteredOptions:
        sortAlphabetically && options ? options.sort() : options || [],
      isOpen: false,
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  clear = () => this._removeOptions();

  focus = () => this.handleAutocompleteClick();

  open = () => this.setState({ isOpen: true });

  close = () => this.setState({ isOpen: false });

  toggleOpen = () => this.setState({ isOpen: !this.state.isOpen });

  handleAutocompleteClick = () => {
    const { inputElement } = this;
    if (inputElement && inputElement.current) {
      inputElement.current.focus();
    }
    // toggle options open/closed
    this.toggleOpen();
  };

  onKeyDown = (event: SyntheticKeyboardEvent<>) => {

    if ( // Check for backspace in order to delete the last selected option
      event.keyCode === 8 &&
      !event.target.value &&
      this.state.selectedOptions.length
    ) {
      // Remove last selected option
      this.removeOption(this.state.selectedOptions.length - 1, event);
    } else if (event.keyCode === 27) { // ESCAPE: prevents Modal from closing
      event.stopPropagation();
    } else if (event.keyCode === 13) { // ENTER: opens Options
      this.open();
    }
  };

  // onChange handler for input element in AutocompleteSkin
  handleInputChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    this._setInputValue(event.target.value);
  };

  // passed to Options onChange handler in AutocompleteSkin
  handleChange = (option: any, event: SyntheticEvent<>) => {
    this.updateSelectedOptions(event, option);
  };

  updateSelectedOptions = (
    event: SyntheticEvent<>,
    selectedOption: any = null
  ) => {
    const canMoreOptionsBeSelected =
      this.state.selectedOptions.length < this.props.maxSelections;
    const areFilteredOptionsAvailable =
      this.state.filteredOptions && this.state.filteredOptions.length > 0;

    if (
      !this.props.maxSelections ||
      (canMoreOptionsBeSelected && areFilteredOptionsAvailable)
    ) {
      if (!selectedOption) return;
      const option = selectedOption.trim();
      const optionCanBeSelected =
        (this.state.selectedOptions.indexOf(option) < 0 &&
          !this.props.multipleSameSelections) ||
        this.props.multipleSameSelections;

      if (option && optionCanBeSelected && this.state.isOpen) {
        const selectedOptions = _.concat(this.state.selectedOptions, option);
        this.selectionChanged(selectedOptions, event);
        this.setState({ selectedOptions, isOpen: false });
      }
    }

    this._setInputValue('');
  };

  removeOption = (index: number, event: SyntheticEvent<>) => {
    const selectedOptions = this.state.selectedOptions;
    _.pullAt(selectedOptions, index);
    this.selectionChanged(selectedOptions, event);
    this.setState({ selectedOptions });
  };

  selectionChanged = (
    selectedOptions: Array<any>,
    event: SyntheticEvent<any>
  ) => {
    if (this.props.onChange) this.props.onChange(selectedOptions, event);
  };

  // returns an object containing props, theme, and method handlers
  // associated with rendering this.state.selectedOptions, the user can call
  // this in the body of the renderSelections function
  getSelectionProps = ({
    removeSelection
  }: { removeSelection: Function } = {}) => {
    const { themeId } = this.props;
    const { inputValue, isOpen, selectedOptions, composedTheme } = this.state;
    return {
      inputValue,
      isOpen,
      selectedOptions,
      theme: composedTheme[themeId],
      removeSelection: (index: number, event: SyntheticEvent<>) =>
        // the user's custom removeSelection event handler is composed with
        // the internal functionality of Autocomplete (this.removeOption)
        composeFunctions(removeSelection, this.removeOption)(index, event)
    };
  };

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      context,
      invalidCharsRegex,
      multipleSameSelections,
      sortAlphabetically,
      skin: AutocompleteSkin,
      theme,
      themeOverrides,
      onChange,
      error,
      ...rest
    } = this.props;

    return (
      <GlobalListeners
        optionsIsOpen={this.state.isOpen}
        optionsRef={this.optionsElement}
        rootRef={this.rootElement}
        toggleOpen={this.toggleOpen}
      >
        {() => (
          <AutocompleteSkin
            error={error || this.state.error}
            filteredOptions={this.state.filteredOptions}
            getSelectionProps={this.getSelectionProps}
            handleAutocompleteClick={this.handleAutocompleteClick}
            handleChange={this.handleChange}
            handleInputChange={this.handleInputChange}
            inputRef={this.inputElement}
            inputValue={this.state.inputValue}
            isOpen={this.state.isOpen}
            onKeyDown={this.onKeyDown}
            optionsRef={this.optionsElement}
            removeOption={this.removeOption}
            rootRef={this.rootElement}
            selectedOptions={this.state.selectedOptions}
            suggestionsRef={this.suggestionsElement}
            theme={this.state.composedTheme}
            toggleOpen={this.toggleOpen}
            {...rest}
          />
        )}
      </GlobalListeners>
    );
  }

  // ======== PRIVATE METHOD ==========

  _removeOptions = () => {
    const { onChange } = this.props;
    onChange ? onChange([]) : null;
    this.setState({ selectedOptions: [], inputValue: '' });
  };

  _filterOptions = (value: string) => {
    let filteredOptions = [];

    if (value !== '') {
      _.some(this.props.options, (option) => {
        if (_.startsWith(option, value)) {
          filteredOptions.push(option);
        }
      });
    } else {
      filteredOptions = this.props.options;
    }

    return filteredOptions;
  };

  _filterInvalidChars = (value: string) => {
    let filteredValue = '';

    if (this.props.invalidCharsRegex.test(value)) {
      filteredValue = value.replace(this.props.invalidCharsRegex, '');
    } else {
      filteredValue = value;
    }

    return filteredValue;
  };

  _setInputValue = (value: string) => {
    const filteredValue = this._filterInvalidChars(value);
    const filteredOptions = this._filterOptions(filteredValue);
    this.setState({
      isOpen: true,
      inputValue: filteredValue,
      filteredOptions
    });
  }
}

export const Autocomplete = withTheme(AutocompleteBase);
