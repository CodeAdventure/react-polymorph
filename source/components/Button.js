// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';

// internal components
import { withTheme } from './HOC/withTheme';

// internal utility functions
import { composeTheme, addThemeId } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '../themes/API';

type Props = {
  className: string,
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  disabled: boolean,
  label: string | Element<any>,
  loading: boolean,
  onClick: Function,
  skin: ComponentType<any>,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object // custom css/scss from user that adheres to component's theme API
};

type State = {
  composedTheme: Object
};

class ButtonBase extends Component<Props, State> {
  // define static properties
  static displayName = 'Button';
  static defaultProps = {
    disabled: false,
    loading: false,
    theme: null,
    themeId: IDENTIFIERS.BUTTON,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      skin: ButtonSkin,
      theme,
      themeOverrides,
      context,
      ...rest
    } = this.props;

    return <ButtonSkin theme={this.state.composedTheme} {...rest} />;
  }
}

export const Button = withTheme(ButtonBase);
