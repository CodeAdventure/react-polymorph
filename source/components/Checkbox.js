import React, { Component } from 'react';
import PropTypes from 'prop-types';

// import utility functions
import { StringOrElement } from '../utils/props.js';
import composeTheme from '../utils/composeTheme.js';

// import the Checkbox component's theme API
import { checkboxThemeAPI } from '../themes/API/checkbox.js';

export default class Checkbox extends Component {
  static propTypes = {
    checked: PropTypes.bool,
    label: StringOrElement,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    skin: PropTypes.func.isRequired,
    theme: PropTypes.object,
    themeOverrides: PropTypes.object,
    themeAPI: PropTypes.object
  };

  static defaultProps = {
    checked: false,
    disabled: false,
    theme: {},
    themeOverrides: {},
    themeAPI: { ...checkboxThemeAPI }
  };

  render() {
    const {
      skin: CheckboxSkin,
      theme,
      themeOverrides,
      themeAPI,
      ...rest
    } = this.props;

    const composedTheme = composeTheme(theme, themeOverrides, themeAPI);

    return <CheckboxSkin theme={composedTheme} {...rest} />;
  }
}
