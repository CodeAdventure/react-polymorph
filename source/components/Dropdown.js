// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';
import createRef from 'create-react-ref/lib/createRef';

// internal utility functions
import { createEmptyContext, withTheme } from './HOC/withTheme';
import { composeTheme, addThemeId, didThemePropsChange } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '.';
import type { ThemeContextProp } from './HOC/withTheme';

type Props = {
  activeItem: any,
  className?: string,
  clickToOpen?: boolean,
  context: ThemeContextProp,
  isOpen?: boolean,
  items: Array<any>,
  label: string | Element<any>,
  noArrow?: boolean,
  onItemSelected?: Function,
  skin?: ComponentType<any>,
  theme: ?Object,
  themeId: string,
  themeOverrides: Object,
};

type State = {
  composedTheme: Object,
  isOpen: boolean,
  isMouseOverItems: boolean,
  isMouseOverRoot: boolean,
};

class DropdownBase extends Component<Props, State> {
  // declare ref types
  rootElement: ?Element<*>;

  // define static properties
  static displayName = 'Dropdown';
  static defaultProps = {
    context: createEmptyContext(),
    clickToOpen: false,
    noArrow: false,
    theme: null,
    themeOverrides: {},
    themeId: IDENTIFIERS.DROPDOWN,
  };

  constructor(props: Props) {
    super(props);

    // define ref
    this.rootElement = createRef();

    const { context, themeId, theme, themeOverrides } = props;
    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      ),
      isOpen: false,
      isMouseOverItems: false,
      isMouseOverRoot: false,
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    didThemePropsChange(this.props, nextProps, this.setState.bind(this));
  }

  // ========= PUBLIC SKIN API =========

  toggleOpen = () => {
    this.setState({ isOpen: !this.state.isOpen });
  };

  // ========= PRIVATE SKIN API =========

  _setMouseOverItems = (isMouseOverItems: boolean) => {
    this.setState({ isMouseOverItems });
  };

  _setMouseOverRoot = (isMouseOverRoot: boolean) => {
    this.setState({ isMouseOverRoot });
  };

  _onItemSelected = (item) => {
    const { onItemSelected } = this.props;
    this._setMouseOverRoot(false);
    this._setMouseOverItems(false);
    this.setState({ isOpen: false });
    if (onItemSelected) {
      onItemSelected(item);
    }
  };

  _onLabelClick = () => {
    this.toggleOpen();
  };

  render() {
    const {
      clickToOpen,
      context,
      isOpen,
      onItemSelected,
      skin,
      theme,
      themeOverrides,
      ...rest
    } = this.props;

    const DropdownSkin = skin || context.skins[IDENTIFIERS.DROPDOWN];
    const { isMouseOverItems, isMouseOverRoot } = this.state;
    const isOpenBecauseOfHover = clickToOpen ? false : isMouseOverItems || isMouseOverRoot;

    return (
      <DropdownSkin
        isOpen={isOpen || this.state.isOpen || isOpenBecauseOfHover}
        onLabelClick={this._onLabelClick}
        onItemSelected={this._onItemSelected}
        rootRef={this.rootElement}
        theme={this.state.composedTheme}
        setMouseOverRoot={this._setMouseOverRoot}
        setMouseOverItems={this._setMouseOverItems}
        {...rest}
      />
    );
  }
}

export const Dropdown = withTheme(DropdownBase);
