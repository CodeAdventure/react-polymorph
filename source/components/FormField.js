// @flow
import React, { Component } from 'react';
import type { ComponentType, Element } from 'react';

// internal components
import { withTheme } from './HOC/withTheme';

// internal utility functions
import { composeTheme, addThemeId } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '../themes/API';

type Props = {
  className: string,
  context: {
    theme: Object,
    ROOT_THEME_API: Object
  },
  disabled: boolean,
  error: string,
  inputRef: Object,
  label: string | Element<any>,
  render: Function,
  skin: ComponentType<any>,
  theme: Object, // will take precedence over theme in context if passed
  themeId: string,
  themeOverrides: Object
};

type State = {
  error: string,
  composedTheme: Object
};

class FormFieldBase extends Component<Props, State> {
  // define static properties
  static displayName = 'FormField';
  static defaultProps = {
    disabled: false,
    theme: null,
    themeId: IDENTIFIERS.FORM_FIELD,
    themeOverrides: {}
  };

  constructor(props: Props) {
    super(props);

    const { context, themeId, theme, themeOverrides } = props;

    this.state = {
      error: '',
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      )
    };
  }

  setError = (error: string) => this.setState({ error });

  focusChild = () => {
    const { inputRef } = this.props;
    if (!inputRef.current) return;
    inputRef.current.focus();
  };

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      skin: FormFieldSkin,
      theme,
      themeOverrides,
      error,
      context,
      inputRef,
      ...rest
    } = this.props;

    return (
      <FormFieldSkin
        error={error || this.state.error}
        setError={this.setError}
        theme={this.state.composedTheme}
        focusChild={this.focusChild}
        {...rest}
      />
    );
  }
}

export const FormField = withTheme(FormFieldBase);
