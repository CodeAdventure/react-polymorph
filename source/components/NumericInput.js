// @flow
import React, { Component } from 'react';
// $FlowFixMe
import type { ComponentType, SyntheticInputEvent, Element, ElementRef } from 'react';

// external libraries
import createRef from 'create-react-ref/lib/createRef';

// internal utility functions
import { createEmptyContext, withTheme } from './HOC/withTheme';
import { composeTheme, addThemeId, didThemePropsChange } from '../utils/themes';

// import constants
import { IDENTIFIERS } from '.';
import type { ThemeContextProp } from './HOC/withTheme';
import { removeCharAtPosition } from '../utils/strings';
import type { InputEvent } from '../utils/types';

type Props = {
  autoFocus?: boolean,
  className?: string,
  context: ThemeContextProp,
  disabled?: boolean,
  error?: string,
  label?: string | Element<any>,
  numberLocaleOptions?: Number$LocaleOptions,
  onBlur?: Function,
  onChange?: Function,
  onFocus?: Function,
  placeholder?: string,
  readOnly?: boolean,
  skin?: ComponentType<any>,
  theme: ?Object,
  themeId: string,
  themeOverrides: Object,
  value: ?number,
};

type State = {
  composedTheme: Object,
  minimumFractionDigits: number,
  inputCaretPosition: number,
  fallbackInputValue: string,
};

// TODO: make this configurable (generalize handling commas and dots in other languages!)
const LOCALE = 'en-US';

class NumericInputBase extends Component<Props, State> {

  inputElement: { current: null | ElementRef<'input'> };

  static displayName = 'NumericInput';

  static defaultProps = {
    context: createEmptyContext(),
    readOnly: false,
    theme: null,
    themeId: IDENTIFIERS.INPUT,
    themeOverrides: {},
    value: null,
  };

  constructor(props: Props) {
    super(props);
    const { context, numberLocaleOptions, themeId, theme, themeOverrides } = props;
    this.inputElement = createRef();
    const minimumFractionDigits = (
      numberLocaleOptions ? numberLocaleOptions.minimumFractionDigits : 0
    );
    this.state = {
      composedTheme: composeTheme(
        addThemeId(theme || context.theme, themeId),
        addThemeId(themeOverrides, themeId),
        context.ROOT_THEME_API
      ),
      minimumFractionDigits: minimumFractionDigits || 0,
      inputCaretPosition: 0,
      fallbackInputValue: '',
    };
  }

  componentDidMount() {
    const { inputElement } = this;
    const { autoFocus } = this.props;
    if (autoFocus) {
      this.focus();
      if (inputElement && inputElement.current) {
        this.setState({
          inputCaretPosition: inputElement.current.selectionStart
        });
      }
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    didThemePropsChange(this.props, nextProps, this.setState.bind(this));
  }

  componentDidUpdate() {
    this.setInputCaretPosition(this.state.inputCaretPosition);
  }

  onChange = (event: SyntheticInputEvent<Element<'input'>>) => {
    event.preventDefault();
    const { value, onChange, disabled } = this.props;
    if (disabled) { return; }
    const result = this.processValueChange(event.nativeEvent);
    if (result) {
      const hasValueChanged = value !== result.value;
      if (hasValueChanged && onChange) {
        onChange(result.value, event);
      }
      this.setState({
        inputCaretPosition: result.caretPosition,
        minimumFractionDigits: result.minimumFractionDigits,
        fallbackInputValue: result.fallbackInputValue,
      });
    }
  };

  /**
   * 1. Handle edge cases that don't need further processing
   * 2. Clean the given value
   * 3. Final processing
   */
  processValueChange(event: InputEvent): ?{
    value: ?number,
    caretPosition: number,
    fallbackInputValue?: string,
    minimumFractionDigits: number,
  } {
    const changedCaretPosition = event.target.selectionStart;
    const valueToProcess = event.target.value;
    const { inputType } = event;
    const { value } = this.props;
    const { fallbackInputValue } = this.state;
    const isBackwardDelete = inputType === 'deleteContentBackward';
    const isForwardDelete = inputType === 'deleteContentForward';
    const deleteCaretCorrection = isBackwardDelete ? 0 : 1;

    /**
     * ========= HANDLE HARD EDGE-CASES =============
     */
    // Case: invalid characters entered -> refuse!
    if (!VALID_INPUT_REGEX.test(valueToProcess)) {
      return {
        caretPosition: changedCaretPosition - 1,
        fallbackInputValue,
        minimumFractionDigits: this.state.minimumFractionDigits,
        value,
      };
    }

    // Case: Everything was deleted -> reset state
    if (valueToProcess === '') {
      return {
        value: null,
        caretPosition: 0,
        fallbackInputValue: '',
        minimumFractionDigits: 0,
      };
    }

    // Case: value is the same as the fallback (which is always shown if defined)
    if (valueToProcess === this.state.fallbackInputValue) return null;

    // Case: Just minus sign was entered
    if (valueToProcess === '-') {
      return {
        value: null,
        caretPosition: 1,
        fallbackInputValue: valueToProcess, // render standalone minus sign
        minimumFractionDigits: 0,
      };
    }

    /**
     * ========= CLEAN THE INPUT =============
     */

    // Options
    const propsMinimumFractionDigits = this.getMinimumFractionDigits();
    const maximumFractionDigits = this.getMaximumFractionDigits();
    const numberLocaleOptions = this.getDynamicLocaleOptions();

    // Current value
    const currentNumber = value;
    const currentValue = (
      value != null ? value.toLocaleString(LOCALE, numberLocaleOptions) : fallbackInputValue
    );
    const currentNumberOfDots = getNumberOfDots(currentValue);
    const hadDotBefore = currentNumberOfDots > 0;

    // New Value
    let newValue = valueToProcess;
    let newCaretPosition = changedCaretPosition;
    const newNumberOfDots = getNumberOfDots(newValue);

    // Case: A second decimal point was added somewhere
    if (hadDotBefore && newNumberOfDots === 2) {
      const oldFirstDotIndex = currentValue.indexOf('.');
      const newFirstDotIndex = newValue.indexOf('.');
      const wasDotAddedBeforeOldOne = newFirstDotIndex < oldFirstDotIndex;
      // Remove the second decimal point and set caret position
      newValue = removeCharAtPosition(
        newValue,
        wasDotAddedBeforeOldOne ? newValue.lastIndexOf('.') : oldFirstDotIndex
      );
      newCaretPosition = newValue.indexOf('.') + 1;
    }

    newValue = truncateToPrecision(newValue, maximumFractionDigits);

    /**
     * ========= PROCESS CLEANED INPUT =============
     */
    const numberOfFractionDigits = getFractionDigits(newValue).length;
    const dynamicMinimumFractionDigits = Math.min(
      Math.max(propsMinimumFractionDigits, numberOfFractionDigits), maximumFractionDigits
    );
    const newNumber = getValueAsNumber(newValue, maximumFractionDigits);

    // Case: Dot was added at the beginning of number

    if (newValue.charAt(0) === '.') {
      return {
        value: null,
        caretPosition: changedCaretPosition,
        fallbackInputValue: newValue, // render new value as-is
        minimumFractionDigits: dynamicMinimumFractionDigits,
      };
    }

    // Case: Invalid change has been made -> ignore it
    if (newNumber == null) {
      const isDeletion = isForwardDelete || isBackwardDelete;
      const deleteAdjustment = isBackwardDelete ? 0 : 1; // special cases when deleting dot
      const insertAdjustment = -1; // don't move caret if numbers are "inserted"
      return {
        caretPosition: changedCaretPosition + (isDeletion ? deleteAdjustment : insertAdjustment),
        fallbackInputValue,
        minimumFractionDigits: dynamicMinimumFractionDigits,
        value: currentNumber,
      };
    }

    // Case: Dot was added at the end of number

    if (newValue.charAt(newValue.length - 1) === '.') {
      return {
        value: null,
        caretPosition: changedCaretPosition,
        fallbackInputValue: newValue, // render new value as-is
        minimumFractionDigits: 0,
      };
    }

    // Case: Dot was deleted with minimum fraction digits constrain defined
    if (this.getMinimumFractionDigitsProp() != null && hadDotBefore && !newNumberOfDots) {
      return {
        caretPosition: newCaretPosition + deleteCaretCorrection,
        fallbackInputValue: '',
        minimumFractionDigits: dynamicMinimumFractionDigits,
        value: currentNumber,
      };
    }

    // Case: Valid change has been made

    const localizedNewNumber = newNumber.toLocaleString(LOCALE, numberLocaleOptions);
    const hasNumberChanged = value !== newNumber;
    const commasDiff = getNumberOfCommas(localizedNewNumber) - getNumberOfCommas(newValue);
    const haveCommasChanged = commasDiff > 0;
    const onlyCommasChanged = !hasNumberChanged && haveCommasChanged;
    const caretCorrection = onlyCommasChanged ? deleteCaretCorrection : commasDiff;
    return {
      caretPosition: Math.max(newCaretPosition + caretCorrection, 0),
      fallbackInputValue: '',
      minimumFractionDigits: dynamicMinimumFractionDigits,
      value: newNumber,
    };
  }

  getMinimumFractionDigitsProp(): number | null {
    const { numberLocaleOptions } = this.props;
    if (!numberLocaleOptions) return null;
    const { minimumFractionDigits } = numberLocaleOptions;
    return minimumFractionDigits || null;
  }

  getMinimumFractionDigits(): number {
    return this.getMinimumFractionDigitsProp() || 0;
  }

  getDynamicMinimumFractionDigits(): number {
    const minimumFractionDigitsProp = this.getMinimumFractionDigits();
    return Math.max(this.state.minimumFractionDigits, minimumFractionDigitsProp || 0);
  }

  getMaximumFractionDigits(): number {
    const { numberLocaleOptions } = this.props;
    const minimumFractionDigits = this.getDynamicMinimumFractionDigits();
    const maximumFractionDigits = (
      numberLocaleOptions && numberLocaleOptions.maximumFractionDigits != null
    ) ? numberLocaleOptions.maximumFractionDigits : 3;
    return Math.max(minimumFractionDigits, maximumFractionDigits);
  }

  getDynamicLocaleOptions(): Number$LocaleOptions {
    return Object.assign({}, this.props.numberLocaleOptions, {
      minimumFractionDigits: this.getDynamicMinimumFractionDigits(),
    });
  }

  getLocalizedNumber(value: ?number) {
    return convertNumberToLocalizedString(value, LOCALE, this.getDynamicLocaleOptions());
  }

  setInputCaretPosition = (position: number) => {
    const { inputElement } = this;
    if (!inputElement.current) return;
    const input = inputElement.current;
    input.selectionStart = position;
    input.selectionEnd = position;
  };

  focus = () => {
    const { inputElement } = this;
    if (!inputElement.current) return;
    inputElement.current.focus();
  };

  render() {
    // destructuring props ensures only the "...rest" get passed down
    const {
      context,
      error,
      numberLocaleOptions,
      onChange,
      skin,
      theme,
      themeOverrides,
      value,
      ...rest
    } = this.props;

    const InputSkin = skin || context.skins[IDENTIFIERS.INPUT];

    const inputValue = value != null ?
      this.getLocalizedNumber(value) :
      this.state.fallbackInputValue;

    return (
      <InputSkin
        error={error}
        inputRef={this.inputElement}
        onChange={this.onChange}
        theme={this.state.composedTheme}
        value={inputValue}
        {...rest}
      />
    );
  }
}

export const NumericInput = withTheme(NumericInputBase);

// ========= HELPERS ==========

const VALID_INPUT_REGEX = /^([0-9,\+\-\.]+)?$/;
const NUMERIC_INPUT_REGEX = /^([\+|\-])?([0-9,]+)?(\.([0-9]+)?)?$/;

const isValidNumericInput = (value: string): boolean => NUMERIC_INPUT_REGEX.test(value);

const isParsableNumberString = (value: string, requiredPrecision: number): boolean => (
  // The number of digits is limited in Javascript - so the required precision influence
  // the possible number of integer digits (only 15 digits can be safely represented in total)
  parseFloat(value) >= (Number.MIN_SAFE_INTEGER / 10 ** (requiredPrecision + 1)) &&
  parseFloat(value) <= (Number.MAX_SAFE_INTEGER / 10 ** (requiredPrecision + 1)) &&
  !isNaN(parseFloat(value)) &&
  isFinite(value)
);

const removeCommas = (value: string): string => value.replace(/,/g, '');

function parseStringToNumber(value: string, requiredPrecision: number): ?number {
  const cleanedValue = removeCommas(value);
  if (!isValidNumericInput(cleanedValue)) return null;
  if (!isParsableNumberString(cleanedValue, requiredPrecision)) return null;
  return parseFloat(cleanedValue);
}

function convertNumberToLocalizedString(
  num: ?number, locale: string, options?: Number$LocaleOptions
): string {
  return num != null ? num.toLocaleString(locale, options) : '';
}

function getValueAsNumber(value: string | number, requiredPrecision: number): ?number {
  return typeof value === 'string' ? parseStringToNumber(value, requiredPrecision) : value;
}

function getNumberOfCommas(value: string): number {
  return (value.match(/,/g) || []).length;
}

function getNumberOfDots(value: string): number {
  return (value.match(/\./g) || []).length;
}

function getIntegerDigits(value: string): string {
  const decimalPointIndex = value.indexOf('.');
  if (decimalPointIndex === -1) return value;
  return value.substring(0, decimalPointIndex);
}

function getFractionDigits(value: string): string {
  const decimalPointIndex = value.indexOf('.');
  if (decimalPointIndex === -1) return '';
  return value.substring(decimalPointIndex + 1);
}

function truncateToPrecision(value: string, precision: number): string {
  const decimalPointIndex = value.indexOf('.');
  if (decimalPointIndex === -1) return value;
  const fractionDigits = removeCommas(getFractionDigits(value));
  return getIntegerDigits(value) + '.' + fractionDigits.substring(0, precision);
}
