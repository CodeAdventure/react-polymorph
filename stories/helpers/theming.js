import React from 'react';
import { ThemeProvider } from '../../source/components/HOC/ThemeProvider';
import { SimpleTheme } from '../../source/themes/simple';

export const decorateWithSimpleTheme = (story) => (
  <ThemeProvider theme={SimpleTheme}>
    {story()}
  </ThemeProvider>
);
