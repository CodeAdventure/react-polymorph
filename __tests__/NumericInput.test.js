import React from 'react';

import { NumericInput } from '../source/components/NumericInput';
import { renderInSimpleTheme } from './helpers/theming';

test('NumericInput renders correctly', () => {
  const component = renderInSimpleTheme(
    <NumericInput />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('NumericInput renders with placeholder', () => {
  const component = renderInSimpleTheme(
    <NumericInput placeholder="0.0000" />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('NumericInput is disabled', () => {
  const component = renderInSimpleTheme(
    <NumericInput disabled />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('NumericInput is readOnly', () => {
  const component = renderInSimpleTheme(
    <NumericInput readOnly />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('NumericInput renders with an error', () => {
  const component = renderInSimpleTheme(
    <NumericInput error="Invalid Amount" />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('NumericInput renders with a value', () => {
  const component = renderInSimpleTheme(
    <NumericInput value="555.333" />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
