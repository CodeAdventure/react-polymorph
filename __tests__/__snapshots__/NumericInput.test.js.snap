// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumericInput is disabled 1`] = `
<div
  className="root disabled"
>
  
  <div
    className="inputWrapper"
  >
    <input
      className="input disabled"
      disabled={true}
      onBlur={[Function]}
      onChange={[Function]}
      readOnly={false}
      value=""
    />
  </div>
</div>
`;

exports[`NumericInput is readOnly 1`] = `
<div
  className="root"
>
  
  <div
    className="inputWrapper"
  >
    <input
      className="input"
      onBlur={[Function]}
      onChange={[Function]}
      readOnly={true}
      value=""
    />
  </div>
</div>
`;

exports[`NumericInput renders correctly 1`] = `
<div
  className="root"
>
  
  <div
    className="inputWrapper"
  >
    <input
      className="input"
      onBlur={[Function]}
      onChange={[Function]}
      readOnly={false}
      value=""
    />
  </div>
</div>
`;

exports[`NumericInput renders with a value 1`] = `
<div
  className="root"
>
  
  <div
    className="inputWrapper"
  >
    <input
      className="input"
      onBlur={[Function]}
      onChange={[Function]}
      readOnly={false}
      value="555.333"
    />
  </div>
</div>
`;

exports[`NumericInput renders with an error 1`] = `
<div
  className="root errored"
>
  <div
    className="error"
  >
    Invalid Amount
  </div>
  <div
    className="inputWrapper"
  >
    <input
      className="input errored"
      onBlur={[Function]}
      onChange={[Function]}
      readOnly={false}
      value=""
    />
  </div>
</div>
`;

exports[`NumericInput renders with placeholder 1`] = `
<div
  className="root"
>
  
  <div
    className="inputWrapper"
  >
    <input
      className="input"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="0.0000"
      readOnly={false}
      value=""
    />
  </div>
</div>
`;
